### PR TITLE
Test script for forceios and forcedroid

### PR DESCRIPTION
**Pull request by wmathurin**

### CODE REVIEW

**Comment by wmathurin** [comment]
I merged test_forceios.js and test_forcedroid.js.
To only review the merge (and some additional changes like improved logging and error handling) see https://github.com/wmathurin/SalesforceMobileSDK-Shared/commit/746bbe5729faa29ea05618749bc0b56230f3056b

**Comment by bhariharan** [comment]
:+1:

File: node/test_force.js [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+}    
+
+//
+// Helper to run arbitrary shell command - errors caught (and reported)
+//
+function runProcessCatchError(cmd, msg) {
+    try {
+        if (msg) log('!TESTING!  ' + msg, COLOR.yellow);
+        log('Running: ' + cmd);
+        execSync(cmd);
```
Comment (khawkins) [inline node/test_force.js]: Are we sure that `execSync` will throw an exception whenever a process exits with a non-zero exit code?  Running this script against iOS, I got errors back:
```
Running: xcodebuild -workspace tmp/testforce855/nativeiosApp/nativeiosApp.xcworkspace -scheme Pods-nativeiosApp clean build
2016-02-04 09:42:07.615 xcodebuild[21345:414749]  DVTAssertions: Warning in /Library/Caches/com.apple.xbs/Sources/DVTFrameworks/DVTFrameworks-9073/DVTFoundation/Utilities/DVTFileSystemUtilities.mm:142
Details:  lstat('/Users/khawkins/Library/Developer/Xcode/DerivedData/nativeiosApp-cpzpeorkqsgtgacffubeatqpqzyf/Build/Products/Debug-iphoneos/SalesforceSDKResources.bundle') failed with errno 2 (No such file or directory)
Function: DVTRecursiveMkdirResult dvt_recursiveMkdir_returningErrorString(NSString *__strong, NSString *__autoreleasing *)
Thread:   <NSThread: 0x7fb76d443dd0>{number = 9, name = (null)}
Please file a bug at http://bugreport.apple.com with this warning message and any useful information you can provide.
```
But the script reported as successful.

Comment (wmathurin) [inline node/test_force.js]: I'll have a look - I'm still working through some issues.

**Comment by wmathurin** [comment]
For unstable on my fork wmathurin, I currently get
```
!SUCCESS! GENERATING native app for android
!SUCCESS! COMPILING native app for android
!SUCCESS! GENERATING react_native app for android
!SUCCESS! COMPILING react_native app for android
!SUCCESS! GENERATING hybrid_local app for android
!FAILURE! COMPILING hybrid_local app for android
!SUCCESS! GENERATING hybrid_remote app for android
!FAILURE! COMPILING hybrid_remote app for android
!SUCCESS! GENERATING native app for ios
!FAILURE! COMPILING native app for ios
!SUCCESS! GENERATING native_swift app for ios
!FAILURE! COMPILING native_swift app for ios
!SUCCESS! GENERATING react_native app for ios
!SUCCESS! COMPILING react_native app for ios
!FAILURE! GENERATING hybrid_local app for ios
!SUCCESS! GENERATING hybrid_remote app for ios
!SUCCESS! COMPILING hybrid_remote app for ios
```

On master, I get
```
!FAILURE! GENERATING native app for android
!FAILURE! GENERATING react_native app for android
!FAILURE! GENERATING hybrid_local app for android
!SUCCESS! GENERATING hybrid_remote app for android
!FAILURE! COMPILING hybrid_remote app for android
!SUCCESS! GENERATING native app for ios
!FAILURE! COMPILING native app for ios
!SUCCESS! GENERATING native_swift app for ios
!FAILURE! COMPILING native_swift app for ios
!SUCCESS! GENERATING react_native app for ios
!SUCCESS! COMPILING react_native app for ios
!FAILURE! GENERATING hybrid_local app for ios
!SUCCESS! GENERATING hybrid_remote app for ios
!SUCCESS! COMPILING hybrid_remote app for ios
```
The android generation failures are due to the shelljs issue (see https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/1043).

I need to investigate the other ones to see if there are legit or due to the script.

**Comment by wmathurin** [comment]
With the shelljs version change for forceios, the failures left on unstable are:
```
!FAILURE! COMPILING hybrid_local app for android
!FAILURE! COMPILING hybrid_remote app for android
!FAILURE! COMPILING native app for ios
!FAILURE! COMPILING native_swift app for ios
```
I'll merge this PR once I resolve them (or figure out that they are not due to a bug in the test script).

**Comment by bhariharan** [comment]
@wmathurin I don't know if this matters, but combining the scripts limits the platform we can run it on to ```OS X```, i.e., we can't test if ```forcedroid``` does the right thing on ```Windows```.

**Comment by wmathurin** [comment]
@bhariharan one of the command line argument for the script is --os=x,y where you pass a comma separated list of platforms you want to test (ios or android). So if you are running on a windows machine, you should only pass android.

**Comment by wmathurin** [comment]
But we should be able to use test_force.js on windows to make sure that all app types for android can be generated and compiled.

File: node/test_force.js [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+            var workspacePath = path.join(appDir, appName + '.xcworkspace');
+            runProcessThrowError('pod update', appDir);    
+            runProcessCatchError('xcodebuild -workspace ' + workspacePath 
+                                 + ' -scheme Pods-' + appName // XXX building Pods-<appName> scheme because <appName> scheme only appears after project is opened in XCode !!
+                                 + ' clean build', 
+                                 'COMPILING ' + target); 
+        }
+        else {
+            // Android - Native
+            runProcessCatchError('./gradlew assembleDebug', 'COMPILING ' + target, appDir);
```
Comment (bhariharan) [inline node/test_force.js]: @wmathurin I think this line will break on ```Windows```. The equivalent is ```gradlew.bat``` on ```Windows``` if I'm not mistaken. I'm not entirely sure though.

Comment (wmathurin) [inline node/test_force.js]: Good catch. I'll add some windows-specific check.

**Comment by wmathurin** [comment]
@bhariharan "windows support" here https://github.com/wmathurin/SalesforceMobileSDK-Shared/commit/592a3ee9c1229938f717477a21448bf2f7c8858e

**Comment by bhariharan** [comment]
:+1:

**Pull request merged**